// @mongodb-js/compass-components/lib/components/modals/modal-header.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModalHeader = exports.Variant = void 0;
const react_1 = __importDefault(require("react"));
const emotion_1 = require("@leafygreen-ui/emotion");
const palette_1 = require("@leafygreen-ui/palette");
const tokens_1 = require("@leafygreen-ui/tokens");
const button_1 = require("@leafygreen-ui/button");
const use_theme_1 = require("../../hooks/use-theme");
const leafygreen_1 = require("../leafygreen");
exports.Variant = {
    Default: button_1.Variant.Primary,
    Danger: button_1.Variant.Danger,
};
const headerStyle = (0, emotion_1.css)({
    padding: tokens_1.spacing[5],
    paddingBottom: 0,
});
const variantStyle = {
    [exports.Variant.Default]: (0, emotion_1.css)({}),
    [exports.Variant.Danger]: (0, emotion_1.css)({
        paddingLeft: '78px',
    }),
};
const titleStyle = (0, emotion_1.css)({
    fontSize: '24px',
    fontWeight: 700,
    lineHeight: '32px',
    margin: 0,
    marginBottom: '10px',
    color: palette_1.palette.black,
});
const titleStyleDark = (0, emotion_1.css)({
    fontWeight: 'bold',
    color: palette_1.palette.gray.light2,
});
const warningIconStyles = (0, emotion_1.css)({
    width: '32px',
    height: '32px',
    borderRadius: '50%',
    background: `${palette_1.palette.red.light3}`,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    left: '36px',
    top: '32px',
    svg: {
        marginTop: '-3px',
    },
});
const warningIconStylesDark = (0, emotion_1.css)({
    background: `${palette_1.palette.red.dark2}`,
});
function ModalHeader({ title, subtitle, variant = exports.Variant.Default, }) {
    const darkMode = (0, use_theme_1.useDarkMode)();
    return (react_1.default.createElement("div", { className: (0, emotion_1.cx)(headerStyle, variantStyle[variant]) },
        variant === exports.Variant.Danger && (react_1.default.createElement("div", { className: (0, emotion_1.cx)(warningIconStyles, darkMode && warningIconStylesDark) },
            react_1.default.createElement(leafygreen_1.Icon, { glyph: "Warning", fill: darkMode ? palette_1.palette.red.light3 : palette_1.palette.red.base, role: "presentation" }))),
        react_1.default.createElement("h1", { className: (0, emotion_1.cx)(titleStyle, darkMode && titleStyleDark), "data-testid": "modal-title", id: "modal-title" }, title),
        subtitle && react_1.default.createElement(leafygreen_1.Body, null, subtitle)));
}
exports.ModalHeader = ModalHeader;
