// @mongodb-js/compass-components/lib/components/horizontal-rule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HorizontalRule = void 0;
const emotion_1 = require("@leafygreen-ui/emotion");
const palette_1 = require("@leafygreen-ui/palette");
const react_1 = __importStar(require("react"));
const use_theme_1 = require("../hooks/use-theme");
const horizontalRuleStyles = (0, emotion_1.css)({
    height: 0,
    width: '100%',
    display: 'block',
    borderTopWidth: 1,
    borderTopStyle: 'solid',
});
const horizontalRuleStylesLight = (0, emotion_1.css)({
    borderTopColor: palette_1.palette.gray.light2,
});
const horizontalRuleStylesDark = (0, emotion_1.css)({
    borderTopColor: palette_1.palette.gray.dark2,
});
exports.HorizontalRule = (0, react_1.forwardRef)(({ className, ...props }, ref) => {
    const darkMode = (0, use_theme_1.useDarkMode)();
    return (react_1.default.createElement("div", { ref: ref, className: (0, emotion_1.cx)(horizontalRuleStyles, darkMode ? horizontalRuleStylesDark : horizontalRuleStylesLight, className), ...props }));
});
exports.HorizontalRule.displayName = 'HorizontalRule';
