// @mongodb-js/compass-components/lib/components/index-keys-badge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexBadge = exports.IndexKeysBadge = void 0;
const leafygreen_1 = require("./leafygreen");
const emotion_1 = require("@leafygreen-ui/emotion");
const tokens_1 = require("@leafygreen-ui/tokens");
const react_1 = __importDefault(require("react"));
const index_icon_1 = __importDefault(require("./index-icon"));
const badge_1 = require("@leafygreen-ui/badge");
const keyListStyles = (0, emotion_1.css)({
    display: 'inline-flex',
    gap: tokens_1.spacing[1],
    marginBottom: tokens_1.spacing[2],
    padding: `0px ${tokens_1.spacing[6]}px`,
});
const badgeStyles = (0, emotion_1.css)({
    textTransform: 'none',
    gap: tokens_1.spacing[1],
});
const IndexBadge = ({ field, value }) => {
    return (react_1.default.createElement(leafygreen_1.Badge, { "data-testid": `${field}-key`, variant: badge_1.Variant.LightGray, className: badgeStyles, role: "listitem" },
        field,
        react_1.default.createElement(index_icon_1.default, { direction: value })));
};
exports.IndexBadge = IndexBadge;
const IndexKeysBadge = ({ keys, 'data-testid': testId, }) => {
    return (react_1.default.createElement("div", { className: keyListStyles, role: "list", "data-testid": testId }, keys.map(({ field, value }) => (react_1.default.createElement(leafygreen_1.Badge, { key: field, "data-testid": `${field}-key`, variant: badge_1.Variant.LightGray, className: badgeStyles, role: "listitem" },
        field,
        react_1.default.createElement(index_icon_1.default, { direction: value }))))));
};
exports.IndexKeysBadge = IndexKeysBadge;
