// @mongodb-js/compass-components/lib/components/leafygreen.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.H2 = exports.H1 = exports.Toggle = exports.TextInput = exports.TextArea = exports.Tabs = exports.Tab = exports.Stage = exports.Cell = exports.Row = exports.TableHead = exports.TableBody = exports.Table = exports.OptionGroup = exports.Option = exports.Select = exports.SegmentedControlOption = exports.SegmentedControl = exports.RadioGroup = exports.Radio = exports.RadioBoxGroup = exports.RadioBox = exports.Popover = exports.Pipeline = exports.MongoDBLogo = exports.MongoDBLogoMark = exports.MarketingModal = exports.ModalFooter = exports.Modal = exports.MenuSeparator = exports.MenuItem = exports.Menu = exports.Language = exports.Icon = exports.HeaderRow = exports.HeaderCell = exports.ExpandedContent = exports.ConfirmationModal = exports.Code = exports.Checkbox = exports.Card = exports.Banner = exports.Badge = exports.AtlasNavGraphic = exports.IconButton = exports.Button = exports.Link = exports.usePrevious = exports.useToast = exports.ToastProvider = void 0;
exports.useLeafyGreenTable = exports.flexRender = exports.InfoSprinkle = exports.SearchInput = exports.Description = exports.Label = exports.Overline = exports.Disclaimer = exports.InlineKeyCode = exports.InlineCode = exports.Body = exports.Subtitle = exports.H3 = void 0;
const react_1 = __importDefault(require("react"));
const badge_1 = __importDefault(require("@leafygreen-ui/badge"));
Object.defineProperty(exports, "Badge", { enumerable: true, get: function () { return badge_1.default; } });
const banner_1 = __importDefault(require("@leafygreen-ui/banner"));
Object.defineProperty(exports, "Banner", { enumerable: true, get: function () { return banner_1.default; } });
const checkbox_1 = __importDefault(require("@leafygreen-ui/checkbox"));
exports.Checkbox = checkbox_1.default;
const card_1 = __importDefault(require("@leafygreen-ui/card"));
exports.Card = card_1.default;
const code_1 = __importStar(require("@leafygreen-ui/code"));
exports.Code = code_1.default;
Object.defineProperty(exports, "Language", { enumerable: true, get: function () { return code_1.Language; } });
const confirmation_modal_1 = __importDefault(require("@leafygreen-ui/confirmation-modal"));
exports.ConfirmationModal = confirmation_modal_1.default;
const icon_1 = __importDefault(require("@leafygreen-ui/icon"));
const logo_1 = require("@leafygreen-ui/logo");
Object.defineProperty(exports, "AtlasNavGraphic", { enumerable: true, get: function () { return logo_1.AtlasNavGraphic; } });
Object.defineProperty(exports, "MongoDBLogoMark", { enumerable: true, get: function () { return logo_1.MongoDBLogoMark; } });
Object.defineProperty(exports, "MongoDBLogo", { enumerable: true, get: function () { return logo_1.MongoDBLogo; } });
const menu_1 = require("@leafygreen-ui/menu");
Object.defineProperty(exports, "MenuSeparator", { enumerable: true, get: function () { return menu_1.MenuSeparator; } });
Object.defineProperty(exports, "MenuItem", { enumerable: true, get: function () { return menu_1.MenuItem; } });
const info_sprinkle_1 = require("@leafygreen-ui/info-sprinkle");
Object.defineProperty(exports, "InfoSprinkle", { enumerable: true, get: function () { return info_sprinkle_1.InfoSprinkle; } });
menu_1.MenuItem.defaultProps = {
    ...menu_1.MenuItem.defaultProps,
    type: 'button',
};
const modal_1 = __importStar(require("@leafygreen-ui/modal"));
exports.Modal = modal_1.default;
Object.defineProperty(exports, "ModalFooter", { enumerable: true, get: function () { return modal_1.Footer; } });
const marketing_modal_1 = __importDefault(require("@leafygreen-ui/marketing-modal"));
exports.MarketingModal = marketing_modal_1.default;
const pipeline_1 = require("@leafygreen-ui/pipeline");
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return pipeline_1.Pipeline; } });
Object.defineProperty(exports, "Stage", { enumerable: true, get: function () { return pipeline_1.Stage; } });
const popover_1 = __importDefault(require("@leafygreen-ui/popover"));
const radio_box_group_1 = require("@leafygreen-ui/radio-box-group");
Object.defineProperty(exports, "RadioBox", { enumerable: true, get: function () { return radio_box_group_1.RadioBox; } });
Object.defineProperty(exports, "RadioBoxGroup", { enumerable: true, get: function () { return radio_box_group_1.RadioBoxGroup; } });
const radio_group_1 = require("@leafygreen-ui/radio-group");
Object.defineProperty(exports, "Radio", { enumerable: true, get: function () { return radio_group_1.Radio; } });
Object.defineProperty(exports, "RadioGroup", { enumerable: true, get: function () { return radio_group_1.RadioGroup; } });
const segmented_control_1 = require("@leafygreen-ui/segmented-control");
Object.defineProperty(exports, "SegmentedControl", { enumerable: true, get: function () { return segmented_control_1.SegmentedControl; } });
Object.defineProperty(exports, "SegmentedControlOption", { enumerable: true, get: function () { return segmented_control_1.SegmentedControlOption; } });
const select_1 = require("@leafygreen-ui/select");
Object.defineProperty(exports, "Select", { enumerable: true, get: function () { return select_1.Select; } });
Object.defineProperty(exports, "Option", { enumerable: true, get: function () { return select_1.Option; } });
Object.defineProperty(exports, "OptionGroup", { enumerable: true, get: function () { return select_1.OptionGroup; } });
const table_1 = require("@leafygreen-ui/table");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return table_1.Cell; } });
Object.defineProperty(exports, "HeaderCell", { enumerable: true, get: function () { return table_1.HeaderCell; } });
Object.defineProperty(exports, "HeaderRow", { enumerable: true, get: function () { return table_1.HeaderRow; } });
Object.defineProperty(exports, "ExpandedContent", { enumerable: true, get: function () { return table_1.ExpandedContent; } });
Object.defineProperty(exports, "Row", { enumerable: true, get: function () { return table_1.Row; } });
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return table_1.Table; } });
Object.defineProperty(exports, "TableHead", { enumerable: true, get: function () { return table_1.TableHead; } });
Object.defineProperty(exports, "TableBody", { enumerable: true, get: function () { return table_1.TableBody; } });
Object.defineProperty(exports, "flexRender", { enumerable: true, get: function () { return table_1.flexRender; } });
Object.defineProperty(exports, "useLeafyGreenTable", { enumerable: true, get: function () { return table_1.useLeafyGreenTable; } });
const tabs_1 = require("@leafygreen-ui/tabs");
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return tabs_1.Tabs; } });
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return tabs_1.Tab; } });
const text_area_1 = __importDefault(require("@leafygreen-ui/text-area"));
exports.TextArea = text_area_1.default;
const text_input_1 = __importDefault(require("@leafygreen-ui/text-input"));
exports.TextInput = text_input_1.default;
const search_input_1 = require("@leafygreen-ui/search-input");
Object.defineProperty(exports, "SearchInput", { enumerable: true, get: function () { return search_input_1.SearchInput; } });
var toast_1 = require("@leafygreen-ui/toast");
Object.defineProperty(exports, "ToastProvider", { enumerable: true, get: function () { return toast_1.ToastProvider; } });
Object.defineProperty(exports, "useToast", { enumerable: true, get: function () { return toast_1.useToast; } });
var hooks_1 = require("@leafygreen-ui/hooks");
Object.defineProperty(exports, "usePrevious", { enumerable: true, get: function () { return hooks_1.usePrevious; } });
const toggle_1 = __importDefault(require("@leafygreen-ui/toggle"));
exports.Toggle = toggle_1.default;
const typography_1 = require("@leafygreen-ui/typography");
Object.defineProperty(exports, "H1", { enumerable: true, get: function () { return typography_1.H1; } });
Object.defineProperty(exports, "H2", { enumerable: true, get: function () { return typography_1.H2; } });
Object.defineProperty(exports, "H3", { enumerable: true, get: function () { return typography_1.H3; } });
Object.defineProperty(exports, "Subtitle", { enumerable: true, get: function () { return typography_1.Subtitle; } });
Object.defineProperty(exports, "Body", { enumerable: true, get: function () { return typography_1.Body; } });
Object.defineProperty(exports, "InlineCode", { enumerable: true, get: function () { return typography_1.InlineCode; } });
Object.defineProperty(exports, "InlineKeyCode", { enumerable: true, get: function () { return typography_1.InlineKeyCode; } });
Object.defineProperty(exports, "Disclaimer", { enumerable: true, get: function () { return typography_1.Disclaimer; } });
Object.defineProperty(exports, "Overline", { enumerable: true, get: function () { return typography_1.Overline; } });
Object.defineProperty(exports, "Label", { enumerable: true, get: function () { return typography_1.Label; } });
Object.defineProperty(exports, "Description", { enumerable: true, get: function () { return typography_1.Description; } });
const use_stacked_component_1 = require("../hooks/use-stacked-component");
const Icon = ({ size, ...rest }) => {
    size = size === 'xsmall' ? 12 : size;
    return react_1.default.createElement(icon_1.default, { size: size, ...rest });
};
exports.Icon = Icon;
Icon.isGlyph = true;
delete marketing_modal_1.default.propTypes;
delete checkbox_1.default.propTypes;
var link_1 = require("./links/link");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return link_1.Link; } });
Object.defineProperty(exports, "Button", { enumerable: true, get: function () { return link_1.Button; } });
Object.defineProperty(exports, "IconButton", { enumerable: true, get: function () { return link_1.IconButton; } });
const Popover = (0, use_stacked_component_1.withStackedComponentPopoverStyles)(popover_1.default);
exports.Popover = Popover;
const Menu = (0, use_stacked_component_1.withStackedComponentPopoverStyles)(menu_1.Menu);
exports.Menu = Menu;
