// @mongodb-js/compass-components/lib/components/list-editor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListEditor = void 0;
const react_1 = __importDefault(require("react"));
const tokens_1 = require("@leafygreen-ui/tokens");
const emotion_1 = require("@leafygreen-ui/emotion");
const leafygreen_1 = require("./leafygreen");
const listEditorStyles = (0, emotion_1.css)({
    display: 'flex',
    flexDirection: 'column',
    gap: tokens_1.spacing[2],
});
const listItemStyles = (0, emotion_1.css)({
    alignItems: 'center',
    display: 'flex',
    gap: tokens_1.spacing[1],
});
const itemContentStyles = (0, emotion_1.css)({
    flexGrow: 1,
});
function ListEditor({ disableAddButton, disableRemoveButton, items, renderItem, onAddItem, onRemoveItem, addButtonTestId, removeButtonTestId, className, itemTestId, itemKey, }) {
    return (react_1.default.createElement("div", { className: (0, emotion_1.cx)(listEditorStyles, className) }, items.map((item, itemIndex) => (react_1.default.createElement("div", { "data-testid": itemTestId?.(itemIndex), className: listItemStyles, key: itemKey?.(item) ?? itemIndex },
        react_1.default.createElement("div", { className: itemContentStyles }, renderItem(item, itemIndex)),
        !disableAddButton?.(item, itemIndex, items) ? (react_1.default.createElement(leafygreen_1.IconButton, { "aria-label": "Add", type: "button", "data-testid": addButtonTestId, disabled: disableAddButton?.(item, itemIndex, items), onClick: () => onAddItem(itemIndex) },
            react_1.default.createElement(leafygreen_1.Icon, { glyph: "Plus" }))) : null,
        !disableRemoveButton?.(item, itemIndex, items) &&
            items.length !== 1 ? (react_1.default.createElement(leafygreen_1.IconButton, { "aria-label": "Remove", type: "button", "data-testid": removeButtonTestId, disabled: disableRemoveButton?.(item, itemIndex, items), onClick: () => onRemoveItem(itemIndex) },
            react_1.default.createElement(leafygreen_1.Icon, { glyph: "Minus" }))) : null)))));
}
exports.ListEditor = ListEditor;
