// @mongodb-js/compass-components/lib/components/toast-body.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToastBody = void 0;
const react_1 = __importDefault(require("react"));
const leafygreen_1 = require("./leafygreen");
const emotion_1 = require("@leafygreen-ui/emotion");
const toastBodyFlexStyles = (0, emotion_1.css)({
    display: 'flex',
    flexDirection: 'row',
});
const toastBodyTextStyles = (0, emotion_1.css)({
    flexGrow: 1,
});
const toastActionStyles = (0, emotion_1.css)({
    textTransform: 'uppercase',
    flexGrow: 0,
    alignSelf: 'center',
    border: 'none',
    padding: 0,
    margin: 0,
    background: 'none',
});
function ToastBody({ statusMessage, actionHandler, actionText, }) {
    return (react_1.default.createElement("div", { className: toastBodyFlexStyles },
        react_1.default.createElement("p", { className: toastBodyTextStyles }, statusMessage),
        !!actionHandler && (react_1.default.createElement(leafygreen_1.Link, { as: "button", "data-testid": `toast-action-${actionText ?? 'none'}`, onClick: actionHandler, hideExternalIcon: true, className: toastActionStyles }, actionText))));
}
exports.ToastBody = ToastBody;
