// @mongodb-js/compass-components/lib/components/document-list/element-actions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddFieldActions = exports.EditActions = void 0;
const react_1 = __importStar(require("react"));
const menu_1 = require("@leafygreen-ui/menu");
const emotion_1 = require("@leafygreen-ui/emotion");
const tokens_1 = require("@leafygreen-ui/tokens");
const leafygreen_1 = require("../leafygreen");
const typography_1 = require("./typography");
const buttonReset = (0, emotion_1.css)({
    margin: 0,
    padding: 0,
    border: 'none',
    background: 'none',
    cursor: 'pointer',
});
const editActionIconStyle = (0, emotion_1.css)({
    position: 'absolute',
    top: 2,
    right: 0,
});
const EditActions = ({ editing, onRemove, onRevert }) => {
    return (react_1.default.createElement(react_1.default.Fragment, null, editing &&
        (onRevert ? (react_1.default.createElement("button", { type: "button", "data-testid": "hadron-document-revert", className: buttonReset, "aria-label": "Revert changes", title: "Revert changes", onClick: (evt) => {
                evt.stopPropagation();
                onRevert();
            } },
            react_1.default.createElement(leafygreen_1.Icon, { size: "xsmall", className: editActionIconStyle, glyph: "Undo" }))) : onRemove ? (react_1.default.createElement("button", { type: "button", "data-testid": "hadron-document-remove", className: buttonReset, title: "Remove field", "aria-label": "Remove field", onClick: (evt) => {
                evt.stopPropagation();
                onRemove();
            } },
            react_1.default.createElement(leafygreen_1.Icon, { size: "xsmall", glyph: "Trash", className: editActionIconStyle }))) : null)));
};
exports.EditActions = EditActions;
const addFieldButton = (0, emotion_1.css)({
    display: 'block',
    width: typography_1.documentTypography.lineHeight,
    height: typography_1.documentTypography.lineHeight,
    marginLeft: 'auto',
    boxShadow: `inset 0 0 0 1px currentColor`,
    borderRadius: '2px',
    userSelect: 'none',
});
const menu = (0, emotion_1.css)({
    width: 'auto',
    minWidth: tokens_1.spacing[6] * 3,
});
const menuItem = (0, emotion_1.css)({
    whiteSpace: 'nowrap',
});
const AddFieldActions = ({ editing, type, parentType, keyName, onAddFieldToElement, onAddFieldAfterElement, }) => {
    const [isOpen, setIsOpen] = (0, react_1.useState)(false);
    if (!editing) {
        return null;
    }
    return (react_1.default.createElement(menu_1.Menu, { open: isOpen, setOpen: setIsOpen, usePortal: false, popoverZIndex: 999999, align: "bottom", justify: "start", className: menu, trigger: ({ children, onClick, ...props }) => {
            return (react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement("button", { type: "button", "data-testid": "hadron-document-add-element", title: "Add field", className: (0, emotion_1.cx)(buttonReset, addFieldButton), onClick: (evt) => {
                        evt.stopPropagation();
                        onClick?.(evt);
                    }, ...props }, "+"),
                children));
        } },
        onAddFieldToElement && (react_1.default.createElement(menu_1.MenuItem, { "data-testid": "hadron-document-add-child", onClick: () => {
                setIsOpen(false);
                onAddFieldToElement();
            }, glyph: react_1.default.createElement(leafygreen_1.Icon, { glyph: "Relationship" }), className: menuItem },
            react_1.default.createElement("div", null,
                "Add ",
                type === 'Array' ? 'item' : 'field',
                " to ",
                react_1.default.createElement("b", null, keyName)))),
        react_1.default.createElement(menu_1.MenuItem, { "data-testid": "hadron-document-add-sibling", onClick: () => {
                setIsOpen(false);
                onAddFieldAfterElement();
            }, glyph: react_1.default.createElement(leafygreen_1.Icon, { glyph: "PlusWithCircle" }), className: menuItem },
            react_1.default.createElement("div", null,
                "Add ",
                parentType === 'Array' ? 'item' : 'field',
                " after ",
                react_1.default.createElement("b", null, keyName)))));
};
exports.AddFieldActions = AddFieldActions;
