// @mongodb-js/compass-components/lib/components/combobox/ComboboxMenu/Menu.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadingIconStyle = exports.menuMessageIconSizeStyle = exports.menuMessageSizeStyle = exports.menuMessageThemeStyle = exports.menuMessageBaseStyle = exports.menuList = exports.menuSizeStyle = exports.menuThemeStyle = exports.menuBaseStyle = exports.popoverThemeStyle = exports.popoverStyle = exports.menuItemHeight = exports.menuItemPadding = void 0;
const polished_1 = require("polished");
const emotion_1 = require("@leafygreen-ui/emotion");
const lib_1 = require("@leafygreen-ui/lib");
const palette_1 = require("@leafygreen-ui/palette");
const tokens_1 = require("@leafygreen-ui/tokens");
const Combobox_types_1 = require("../Combobox.types");
exports.menuItemPadding = {
    [Combobox_types_1.ComboboxSize.Default]: { x: 12, y: 8 },
    [Combobox_types_1.ComboboxSize.Large]: { x: 12, y: 8 },
};
exports.menuItemHeight = {
    [Combobox_types_1.ComboboxSize.Default]: tokens_1.typeScales.body1.lineHeight + 2 * exports.menuItemPadding[Combobox_types_1.ComboboxSize.Default].y,
    [Combobox_types_1.ComboboxSize.Large]: tokens_1.typeScales.body2.lineHeight + 2 * exports.menuItemPadding[Combobox_types_1.ComboboxSize.Large].y,
};
exports.popoverStyle = (0, emotion_1.css) `
  overflow: hidden;
  border-radius: 12px;
  border: 1px solid;
`;
exports.popoverThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    box-shadow: 0px 4px 7px ${(0, polished_1.transparentize)(0.85, palette_1.palette.black)};
    border-color: ${palette_1.palette.gray.light2};
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) `
    box-shadow: 0px 4px 7px ${(0, polished_1.transparentize)(0.85, palette_1.palette.black)};
    border-color: ${palette_1.palette.gray.dark3};
  `,
};
exports.menuBaseStyle = (0, emotion_1.css) `
  position: relative;
  width: 100%;
  margin: 0;
  padding: ${tokens_1.spacing[2]}px 0;
  font-family: ${tokens_1.fontFamilies.default};
  border-radius: inherit;
  overflow-y: auto;
  scroll-behavior: smooth;
`;
exports.menuThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    color: ${palette_1.palette.black};
    background-color: ${palette_1.palette.white};
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) `
    color: ${palette_1.palette.gray.light1};
    background-color: ${palette_1.palette.gray.dark3};
  `,
};
exports.menuSizeStyle = {
    [Combobox_types_1.ComboboxSize.Default]: (0, emotion_1.css) `
    min-height: ${exports.menuItemHeight[Combobox_types_1.ComboboxSize.Default]}px;
  `,
    [Combobox_types_1.ComboboxSize.Large]: (0, emotion_1.css) `
    min-height: ${exports.menuItemHeight[Combobox_types_1.ComboboxSize.Large]}px;
  `,
};
exports.menuList = (0, emotion_1.css) `
  position: relative;
  margin: 0;
  padding: 0;
`;
exports.menuMessageBaseStyle = (0, emotion_1.css) `
  font-family: inherit;
  display: inline-flex;
  align-items: center;
  gap: 8px;
`;
exports.menuMessageThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    color: ${palette_1.palette.gray.dark3};
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) `
    color: ${palette_1.palette.gray.light3};
  `,
};
exports.menuMessageSizeStyle = {
    [Combobox_types_1.ComboboxSize.Default]: (0, emotion_1.css) `
    font-size: ${tokens_1.typeScales.body1.fontSize}px;
    line-height: ${tokens_1.typeScales.body1.lineHeight}px;
    padding: ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Default].y}px
      ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Default].x}px;
  `,
    [Combobox_types_1.ComboboxSize.Large]: (0, emotion_1.css) `
    font-size: ${tokens_1.typeScales.body2.fontSize}px;
    line-height: ${tokens_1.typeScales.body2.lineHeight}px;
    padding: ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Large].y}px
      ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Large].x}px;
  `,
};
exports.menuMessageIconSizeStyle = {
    [Combobox_types_1.ComboboxSize.Default]: (0, emotion_1.css) `
    height: ${tokens_1.typeScales.body1.fontSize}px;
    width: ${tokens_1.typeScales.body1.fontSize}px;
  `,
    [Combobox_types_1.ComboboxSize.Large]: (0, emotion_1.css) `
    height: ${tokens_1.typeScales.body2.fontSize}px;
    width: ${tokens_1.typeScales.body2.fontSize}px;
  `,
};
const loadingIconAnimation = (0, emotion_1.keyframes) `
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`;
exports.loadingIconStyle = (0, emotion_1.css) `
  animation: ${loadingIconAnimation} 1.5s linear infinite;
`;
