// hadron-document/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.objectToIdiomaticEJSON = exports.getDefaultValueForType = exports.isInternalFieldPath = exports.ElementEditor = exports.ElementEvents = exports.Element = exports.DocumentEvents = exports.Document = void 0;
const document_1 = __importStar(require("./document"));
exports.Document = document_1.default;
Object.defineProperty(exports, "DocumentEvents", { enumerable: true, get: function () { return document_1.Events; } });
const element_1 = __importStar(require("./element"));
exports.Element = element_1.default;
Object.defineProperty(exports, "ElementEvents", { enumerable: true, get: function () { return element_1.Events; } });
Object.defineProperty(exports, "isInternalFieldPath", { enumerable: true, get: function () { return element_1.isInternalFieldPath; } });
const editor_1 = __importDefault(require("./editor"));
exports.ElementEditor = editor_1.default;
const utils_1 = require("./utils");
Object.defineProperty(exports, "getDefaultValueForType", { enumerable: true, get: function () { return utils_1.getDefaultValueForType; } });
Object.defineProperty(exports, "objectToIdiomaticEJSON", { enumerable: true, get: function () { return utils_1.objectToIdiomaticEJSON; } });
exports.default = document_1.default;
