// @mongodb-js/compass-components/lib/components/combobox/Combobox.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNullSelection = exports.SearchState = exports.State = exports.Overflow = exports.TruncationLocation = exports.ComboboxSize = exports.ComboboxElement = void 0;
exports.ComboboxElement = {
    Input: 'Input',
    ClearButton: 'ClearButton',
    FirstChip: 'FirstChip',
    LastChip: 'LastChip',
    MiddleChip: 'MiddleChip',
    Combobox: 'Combobox',
    Menu: 'Menu',
};
exports.ComboboxSize = {
    Default: 'default',
    Large: 'large',
};
exports.TruncationLocation = {
    start: 'start',
    middle: 'middle',
    end: 'end',
    none: 'none',
};
exports.Overflow = {
    expandY: 'expand-y',
    scrollX: 'scroll-x',
    expandX: 'expand-x',
};
exports.State = {
    error: 'error',
    none: 'none',
};
exports.SearchState = {
    unset: 'unset',
    error: 'error',
    loading: 'loading',
};
function getNullSelection(multiselect) {
    if (multiselect) {
        return [];
    }
    else {
        return null;
    }
}
exports.getNullSelection = getNullSelection;
