// @mongodb-js/compass-components/lib/components/combobox/Menu.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadingIconStyle = exports.menuMessageIconSizeStyle = exports.menuMessageSizeStyle = exports.menuMessageThemeStyle = exports.menuMessageBaseStyle = exports.menuList = exports.menuThemeStyle = exports.menuBaseStyle = exports.popoverThemeStyle = exports.popoverStyle = exports.menuItemPadding = exports.menuItemHeight = void 0;
const emotion_1 = require("@leafygreen-ui/emotion");
const lib_1 = require("@leafygreen-ui/lib");
const palette_1 = require("@leafygreen-ui/palette");
const tokens_1 = require("@leafygreen-ui/tokens");
const Combobox_types_1 = require("./Combobox.types");
exports.menuItemHeight = 36;
exports.menuItemPadding = {
    [Combobox_types_1.ComboboxSize.Default]: { x: 12, y: 8 },
    [Combobox_types_1.ComboboxSize.Large]: { x: 12, y: 8 },
};
const popoverStyle = (width = 384) => (0, emotion_1.css) `
  border-radius: 4px;
  width: ${width}px;
  overflow: hidden;
`;
exports.popoverStyle = popoverStyle;
exports.popoverThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    box-shadow: 0px 3px 7px rgba(0, 0, 0, 0.25);
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) ``,
};
exports.menuBaseStyle = (0, emotion_1.css) `
  position: relative;
  width: 100%;
  margin: 0;
  padding: 0;
  font-family: ${tokens_1.fontFamilies.default};
  border-radius: inherit;
  overflow-y: auto;
  scroll-behavior: smooth;
  min-height: ${exports.menuItemHeight}px;
`;
exports.menuThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    color: ${palette_1.palette.gray.dark3};
    background-color: ${palette_1.palette.white};
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) ``,
};
exports.menuList = (0, emotion_1.css) `
  position: relative;
  margin: 0;
  padding: 0;
`;
exports.menuMessageBaseStyle = (0, emotion_1.css) `
  font-family: inherit;
  display: inline-flex;
  align-items: center;
  gap: 8px;
`;
exports.menuMessageThemeStyle = {
    [lib_1.Theme.Light]: (0, emotion_1.css) `
    ${palette_1.palette.gray.dark1}
  `,
    [lib_1.Theme.Dark]: (0, emotion_1.css) ``,
};
exports.menuMessageSizeStyle = {
    [Combobox_types_1.ComboboxSize.Default]: (0, emotion_1.css) `
    font-size: ${tokens_1.typeScales.body1.fontSize +
        1}px; // TODO: update this @ redesign
    line-height: ${tokens_1.typeScales.body1.lineHeight +
        1}px; // TODO: update this @ redesign
    padding: ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Default].y}px
      ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Default].x}px;
  `,
    [Combobox_types_1.ComboboxSize.Large]: (0, emotion_1.css) `
    font-size: ${tokens_1.typeScales.body2.fontSize}px;
    line-height: ${tokens_1.typeScales.body2.lineHeight}px;
    padding: ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Large].y}px
      ${exports.menuItemPadding[Combobox_types_1.ComboboxSize.Large].x}px;
  `,
};
exports.menuMessageIconSizeStyle = {
    [Combobox_types_1.ComboboxSize.Default]: (0, emotion_1.css) `
    height: ${tokens_1.typeScales.body1.fontSize + 1}px; // TODO: update this @ redesign
    width: ${tokens_1.typeScales.body1.fontSize + 1}px; // TODO: update this @ redesign
  `,
    [Combobox_types_1.ComboboxSize.Large]: (0, emotion_1.css) `
    height: ${tokens_1.typeScales.body2.fontSize}px;
    width: ${tokens_1.typeScales.body2.fontSize}px;
  `,
};
const loadingIconAnimation = (0, emotion_1.keyframes) `
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`;
exports.loadingIconStyle = (0, emotion_1.css) `
  animation: ${loadingIconAnimation} 1.5s linear infinite;
`;
