// @mongodb-js/compass-components/lib/utils/format-date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatDate = void 0;
const SECOND = 1000;
const MINUTE = SECOND * 60;
const HOUR = MINUTE * 60;
const DAY = HOUR * 24;
const WEEK = DAY * 7;
const MONTH = WEEK * 4;
const relativeDateFormatter = new Intl.RelativeTimeFormat('en', {
    numeric: 'auto',
});
const absoluteDateFormatter = new Intl.DateTimeFormat('en', {
    dateStyle: 'long',
});
const formatScale = {
    second: SECOND,
    minute: MINUTE,
    hour: HOUR,
    day: DAY,
    week: WEEK,
};
function formatDate(date) {
    const timeDiff = Date.now() - date;
    let formatType = null;
    if (timeDiff < MINUTE) {
        formatType = 'second';
    }
    else if (timeDiff < HOUR) {
        formatType = 'minute';
    }
    else if (timeDiff < DAY) {
        formatType = 'hour';
    }
    else if (timeDiff < WEEK) {
        formatType = 'day';
    }
    else if (timeDiff < MONTH) {
        formatType = 'week';
    }
    else {
        return absoluteDateFormatter.format(date);
    }
    return relativeDateFormatter.format(-Math.round(timeDiff / formatScale[formatType]), formatType);
}
exports.formatDate = formatDate;
