// hadron-document/dist/editor/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectIdEditor = exports.UndefinedEditor = exports.NullEditor = exports.Int64Editor = exports.Int32Editor = exports.DoubleEditor = exports.Decimal128Editor = exports.StringEditor = exports.StandardEditor = exports.DateEditor = void 0;
const standard_1 = __importDefault(require("./standard"));
exports.StandardEditor = standard_1.default;
const string_1 = __importDefault(require("./string"));
exports.StringEditor = string_1.default;
const decimal128_1 = __importDefault(require("./decimal128"));
exports.Decimal128Editor = decimal128_1.default;
const int32_1 = __importDefault(require("./int32"));
exports.Int32Editor = int32_1.default;
const int64_1 = __importDefault(require("./int64"));
exports.Int64Editor = int64_1.default;
const double_1 = __importDefault(require("./double"));
exports.DoubleEditor = double_1.default;
const date_1 = __importDefault(require("./date"));
exports.DateEditor = date_1.default;
const null_1 = __importDefault(require("./null"));
exports.NullEditor = null_1.default;
const undefined_1 = __importDefault(require("./undefined"));
exports.UndefinedEditor = undefined_1.default;
const objectid_1 = __importDefault(require("./objectid"));
exports.ObjectIdEditor = objectid_1.default;
const init = (element) => ({
    Standard: new standard_1.default(element),
    String: new string_1.default(element),
    Decimal128: new decimal128_1.default(element),
    Date: new date_1.default(element),
    Double: new double_1.default(element),
    Int32: new int32_1.default(element),
    Int64: new int64_1.default(element),
    Null: new null_1.default(element),
    Undefined: new undefined_1.default(element),
    ObjectId: new objectid_1.default(element),
});
exports.default = Object.assign(init, {
    DateEditor: date_1.default,
    StandardEditor: standard_1.default,
    StringEditor: string_1.default,
    Decimal128Editor: decimal128_1.default,
    DoubleEditor: double_1.default,
    Int32Editor: int32_1.default,
    Int64Editor: int64_1.default,
    NullEditor: null_1.default,
    UndefinedEditor: undefined_1.default,
    ObjectIdEditor: objectid_1.default,
});
