// hadron-document/dist/editor/standard.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const hadron_type_checker_1 = __importDefault(require("hadron-type-checker"));
const element_events_1 = __importDefault(require("../element-events"));
const utils_1 = require("../utils");
const ARRAY_OR_OBJECT = /^(\[|\{)(.+)(\]|\})$/;
class StandardEditor {
    constructor(element) {
        this.element = element;
        this.type = element.currentType;
        this.editing = false;
    }
    edit(value) {
        const currentType = this.element.currentType;
        try {
            const newValue = hadron_type_checker_1.default.cast(value, currentType);
            this.element.edit(newValue);
        }
        catch (e) {
            this.element.setInvalid(value, currentType, e.message);
        }
    }
    paste(value) {
        if (ARRAY_OR_OBJECT.exec(value)) {
            this.edit(JSON.parse(value));
            this.element._bubbleUp(element_events_1.default.Converted, this.element);
        }
        else {
            this.edit(value);
        }
    }
    size() {
        return (0, utils_1.fieldStringLen)(this.element.currentValue);
    }
    value() {
        return String(this.element.currentValue);
    }
    start() {
        this.editing = true;
    }
    complete() {
        this.editing = false;
    }
}
exports.default = StandardEditor;
